stdlib/public/CMakeLists.txt

```diff
@@ -190,6 +190,12 @@ if(SWIFT_SHOULD_BUILD_EMBEDDED_STDLIB)
       "riscv64  riscv64-none-none-eabi    riscv64-none-none-eabi"
     )
   endif()
+  if("X86" IN_LIST LLVM_TARGETS_TO_BUILD)
+    list(APPEND EMBEDDED_STDLIB_TARGET_TRIPLES
+      "i686    i686-none-none-eabi       i686-none-none-eabi"
+      "x86_64  x86_64-none-none-eabi     x86_64-none-none-eabi"
```

Comment:
There is no EABI specifications for these platforms.  I think that the target triples used here should be `i686-unknown-none-elf`, `i686-unknown-none-coff` and their respective `x86_64-unknown-none-elf`, and `x86_64-unknown-none-coff`.

Reply:
This fails to build the stdlib though with this error, I guess LLVM doesn't support the coff triples that you named?

```
<unknown>:0: error: fatal error encountered during compilation; please submit a bug report (https://swift.org/contributing/#reporting-bugs)
<unknown>:0: note: Cannot initialize MC for non-Windows COFF object files.
Please submit a bug report (https://swift.org/contributing/#reporting-bugs) and include the crash backtrace.
Stack dump:
0.	<eof> parser at end of file
Stack dump without symbol names (ensure you have llvm-symbolizer in your PATH or set the environment var `LLVM_SYMBOLIZER_PATH` to point to it):
0  swift-frontend           0x00000001073587e7 llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) + 39
1  swift-frontend           0x0000000107356cd5 llvm::sys::RunSignalHandlers() + 85
2  swift-frontend           0x0000000107358e80 SignalHandler(int) + 288
3  libsystem_platform.dylib 0x00007ff803da5dfd _sigtramp + 29
4  libsystem_platform.dylib 0x00007fd78d818a00 _sigtramp + 18446743934285065248
5  libsystem_c.dylib        0x00007ff803cdbd24 abort + 123
6  swift-frontend           0x00000001012423f9 swift::performFrontend(llvm::ArrayRef<char const*>, char const*, void*, swift::FrontendObserver*)::$_3::__invoke(void*, char const*, bool) + 585
7  swift-frontend           0x00000001072c60ed llvm::report_fatal_error(llvm::Twine const&, bool) + 365
8  swift-frontend           0x00000001072c5f79 llvm::report_fatal_error(char const*, bool) + 41
9  swift-frontend           0x0000000106e0143b llvm::MCContext::MCContext(llvm::Triple const&, llvm::MCAsmInfo const*, llvm::MCRegisterInfo const*, llvm::MCSubtargetInfo const*, llvm::SourceMgr const*, llvm::MCTargetOptions const*, bool, llvm::StringRef) + 1851
10 swift-frontend           0x0000000104bc6da4 llvm::MachineModuleInfoWrapperPass::MachineModuleInfoWrapperPass(llvm::LLVMTargetMachine const*) + 132
11 swift-frontend           0x0000000104b34bc7 llvm::LLVMTargetMachine::addPassesToEmitFile(llvm::legacy::PassManagerBase&, llvm::raw_pwrite_stream&, llvm::raw_pwrite_stream*, llvm::CodeGenFileType, bool, llvm::MachineModuleInfoWrapperPass*, llvm::raw_pwrite_stream*) + 71
12 swift-frontend           0x00000001040fb9c9 clang::EmitBackendOutput(clang::DiagnosticsEngine&, clang::HeaderSearchOptions const&, clang::CodeGenOptions const&, clang::TargetOptions const&, clang::LangOptions const&, clang::CASOptions const&, llvm::StringRef, llvm::Module*, clang::BackendAction, llvm::IntrusiveRefCntPtr<llvm::vfs::FileSystem>, std::__1::unique_ptr<llvm::raw_pwrite_stream, std::__1::default_delete<llvm::raw_pwrite_stream>>, std::__1::unique_ptr<llvm::raw_pwrite_stream, std::__1::default_delete<llvm::raw_pwrite_stream>>) + 4345
13 swift-frontend           0x00000001045d0ccb (anonymous namespace)::PCHContainerGenerator::HandleTranslationUnit(clang::ASTContext&) + 2043
14 swift-frontend           0x00000001050fe80c clang::MultiplexConsumer::HandleTranslationUnit(clang::ASTContext&) + 44
15 swift-frontend           0x00000001052cef95 clang::ParseAST(clang::Sema&, bool, bool) + 597
16 swift-frontend           0x00000001050cb414 clang::FrontendAction::Execute() + 84
17 swift-frontend           0x000000010504b465 clang::CompilerInstance::ExecuteAction(clang::FrontendAction&) + 757
18 swift-frontend           0x0000000105057d92 void llvm::function_ref<void ()>::callback_fn<compileModuleImpl(clang::CompilerInstance&, clang::SourceLocation, llvm::StringRef, clang::FrontendInputFile, llvm::StringRef, llvm::StringRef, llvm::function_ref<void (clang::CompilerInstance&)>, llvm::function_ref<void (clang::CompilerInstance&)>)::$_8>(long) + 130
19 swift-frontend           0x00000001072bcafe llvm::CrashRecoveryContext::RunSafely(llvm::function_ref<void ()>) + 126
20 swift-frontend           0x00000001072bcc00 RunSafelyOnThread_Dispatch(void*) + 48
21 swift-frontend           0x00000001072bcc1f void* llvm::thread::ThreadProxy<std::__1::tuple<void (*)(void*), (anonymous namespace)::RunSafelyOnThreadInfo*>>(void*) + 15
22 libsystem_pthread.dylib  0x00007ff803d904e1 _pthread_start + 125
23 libsystem_pthread.dylib  0x00007ff803d8bf6b thread_start + 15
```

Reply:
Ah, interesting.  Okay; I guess for now we can leave that off.  It is unfortunate, as that does preclude a few embedded use cases (e.g. the possibility of replacing TianoCore with something written in a memory safe language).

Reply:
I still want to be able to emit valid PE files in the embedded. Would it make sense to try adding `i686-pc-windows-coff` and `x86_64-pc-windows-coff` to this list instead?

Reply:
If it is specifically for Windows, I think that might be possible as a workaround. The problem is that there are places that will key off the OS, and I am concerned that it will break the truly embedded use cases (e.g. UEFI).

Reply:
UEFI uses PE-COFF though (like Windows), as opposed to vanilla COFF, so we should probably have support for `x86_64-pc-uefi-coff` or something along those lines for that one.

Reply:
I disagree on that naming.

1. The `pc` vendor is unnecessary, I think that `unknown` would be fine for the vendor.
2. COFF is the right triple component for the *object file format*. PE is not an object file format, it is the linked format and that is not represented in the triple. This would be similar to the `x86_64-apple-none-macho` triple used for the UEFI stub along with mtoc.

I think that x86_64-unknown-coff is a pretty useful and interesting triple for bare metal and UEFI. I think that PE/COFF gives you a much better format than ELF, but we can have different opinions :)

Reply:
> PE is not an object file format, it is the linked format and that is not represented in the triple.

IMO the triples represent _both_ the object format _and_ the final linked format, hence `x86_64-<whatever>-windows-coff` causes the compiler to generate PE-COFF executables. UNIX systems just happen to use the same format for the executable as they do for object files (well, give or take; the name is the same, at least, even if there are differences in the detail), while many other systems don't, but I don't believe that means the triple only tells you one or the other.

> I think that x86_64-unknown-coff is a pretty useful and interesting triple for bare metal and UEFI.

For bare metal it's arguably useful, and indeed in the past I _have_ used `m68k-unknown-coff` and `powerpc-unknown-coff` in custom toolchains for that reason. I think these days I might well choose ELF instead, though I acknowledge there might be reasons to prefer COFF or even a.out in some contexts.

For UEFI though, the final image needs to be PE-COFF, so I think we need the `uefi` environment designator.

> I think that PE/COFF gives you a much better format than ELF, but we can have different opinions :)

:-) I'm going to avoid being drawn on that.

Reply:
> IMO the triples represent _both_ the object format _and_ the final linked format, hence `x86_64-<whatever>-windows-coff` causes the compiler to generate PE-COFF executables. UNIX systems just happen to use the same format for the executable as they do for object files (well, give or take; the name is the same, at least, even if there are differences in the detail), while many other systems don't, but I don't believe that means the triple only tells you one or the other.

I don't think that is the case. Windows is the OS, you _can_ use a different OS and still use PE/COFF. The linker generates the final executable from object files, and the compiler emits object files. We can make certain assumptions knowing the OS, but not knowing the OS doesn't really preclude the linker from generating something else. I don't think that the compiler emits anything PE specific in the COFF object file itself.

> For UEFI though, the final image needs to be PE-COFF, so I think we need the uefi environment designator.

Is there a UEFI environment designator in config.sub in the GNU project and the LLVM triple type? The latter is modelled after the former.

Reply:
> I don't think that is the case. Windows is the OS, you can use a different OS and still use PE/COFF.

Agreed. But if you were building for e.g. ReactOS, you would also need to use a triple that caused the toolchain to use a PE/COFF linker.

Put another way, if you run `x86_64-unknown-windows-coff-gcc`, that will execute a linker called `x86_64-unknown-windows-coff-ld` (or similar), no? That is, it will choose a linker based on the triple, at least by default, unless there's some special logic in GCC or an override in the `.specs` file for that triple. (And in both of those cases, it's still the triple that will determine which linker to use, and the linker is, I think we both agree, what determines the final output format.)

> Is there a UEFI environment designator in config.sub in the GNU project and the LLVM triple type? The latter is modelled after the former.

Good question. I don't have a GCC source tree handy to check, but I don't see it in [config.sub](https://github.com/gcc-mirror/gcc/blob/master/config.sub). Does anyone use GCC to build EFI binaries? I've always imagined that people use MSVC and/or ICC for those, given the choice of PE/COFF and Intel architecture.

Reply:
> Put another way, if you run x86_64-unknown-windows-coff-gcc, that will execute a linker called x86_64-unknown-windows-coff-ld (or similar), no? That is, it will choose a linker based on the triple, at least by default, unless there's some special logic in GCC or an override in the .specs file for that triple. (And in both of those cases, it's still the triple that will determine which linker to use, and the linker is, I think we both agree, what determines the final output format.)

That used to be the default, but not so much anymore. It will fall back to `ld` which will fail. But you also now have `-fuse-ld=...` so override that irrespective of the specs. With `ld`, the assumption was that it is BFD, and you can use `-Xlinker -m,x86_64-pecoff` to select the correct emulation mode as well. So, you would use `x86_64-unknown-windows-coff-gcc -Xlinker -m,x86_64-pecoff` which would use `ld.bfd` and you a PE/COFF binary on whatever host you are running. Yes, we both agree that the linker controls the final output format, however, the triple does not control the linker anymore with the introduction of `-fuse-ld=`.

> I don't have a GCC source tree handy to check, but I don't see it in [config.sub](https://github.com/gcc-mirror/gcc/blob/master/config.sub).

Which leads me to believe that `uefi` is not a well-formed environment.